Automatically compile locales from django.po to django.mo when building.

allo- opened the pull request:
fixes #188

joshblum commented on setup.py:
```diff
@@ -17,6 +20,35 @@ def run_tests(self):
+    def run(self):
+        _install_lib.run(self)
+        try:
+            # django before 1.7
```
given #191 we can remove this

joshblum replied:
@allo- can you update to remove the django 1.7 changes since we no longer support this version?

allo- commented:
If you really want this ... because it should not hurt on newer versions and people still using 1.7 could use it.

joshblum commented:
@allo- we explicitly dropped django 1.7 support so it would be better to fail loudly in my opinion

allo- commented:
Maybe there should be an explicit check? On the other hand ,the requirements.txt will produce a conflict for wrong django anyway. I added some comment on tested versions and how it works.

Not sure how stable the api for compile messages is. But as using the ``.handle()`` method corresponds to immitating the command line, it will probably work as long as command handling in django doesn't change.

joshblum commented on setup.py:
```diff
@@ -17,6 +20,26 @@ def run_tests(self):
+            from django.core.management.commands.compilemessages import Command
+            os.chdir('registration')
+            cmd = Command()
+            cmd.handle(verbosity=0, exclude=[], locale=[])
```
can you use the `call_command` api? https://docs.djangoproject.com/en/1.9/ref/django-admin/#running-management-commands-from-your-code

allo- commented:
I guess thats the cleaner way. Need to look into it, when i have the time to test it.

joshblum commented:
@allo- any updates?

allo- commented:
Is there a reasonble option for testing the install in the way, that a install should create .mo files? This might be a useful unit test.

joshblum commented on setup.py:
```diff
@@ -17,6 +20,25 @@ def run_tests(self):
+            os.chdir('registration')
+            call_command("compilemessages")
+            os.chdir('..')
+            return
```
remove

joshblum replied:
this line is not needed

joshblum commented on setup.py:
```diff
@@ -17,6 +20,25 @@ def run_tests(self):
+        from the .po source files in .mo files.
+    """
+    def run(self):
+        _install_lib.run(self)
```
`super()`

joshblum commented on setup.py:
```diff
@@ -17,6 +20,25 @@ def run_tests(self):
+        _install_lib.run(self)
+        try:
+            # Use the management command to compile messages
+            # tested with django 1.7-1.9, 1.9 does not need the chdir anymore
```
1.8

joshblum replied:
we no longer support version 1.7, so this comment should just read from 1.8-1.9

joshblum commented on setup.py:
```diff
@@ -17,6 +20,25 @@ def run_tests(self):
+        try:
+            # Use the management command to compile messages
+            # tested with django 1.7-1.9, 1.9 does not need the chdir anymore
+            from django.core.management import call_command
```
why not import at the top of the file?

joshblum replied:
we only support django version 1.8+, so it shouldn't be problem right?

joshblum commented:
@allo- a few small nits, then lgtm. @dicato any ideas for testing or should we leave as is?

allo- commented:
- "remove": I do not understand what you want to remove
- "super": Yes, should be changed.
- "1.8"? I tested this part of the code from 1.7 to 1.9 and it ran without problems. Or what do you mean?
- "import not on top": This allows to fail graceful when the import fails (i.e. wrong django version). Before i even had a setup.py checked in, which works with older versions as well.

allo- commented:
I find it kind of hard to follow the thread, when your answers appear above at the line notes.

- So which line to remove? You're quoting the whole class there.
- I would keep the comment, as it's definitely correct. Of course we could change it to 1.8-1.9 but what's the gain?
- "we only support django version 1.8+": Which line shouldn't be the problem? The conditional import?

joshblum commented:
sorry for the confusion, the lines I commented were the relevant ones. 

- the `return` at L38 is not needed
- The comment is fine, it should just read 1.8-1.9. We don't support 1.7 so there is no reason to mention it
- I think that the import can be moved to the top of the file since we only support 1.8+, we shouldn't have any import errors

dicato commented:
@joshblum I think the only way to test is to try a few supported locales and ensure the templates have the correct messages.

allo- commented:
removed the "tested with" completely, as it runs with the currently supported versions. This was only interesting, if a wide range of versions should be supported. The return was left over from before where the next code block tried the old method.

joshblum commented:
@allo- thanks for making these small changes, can you also change L31 to call `super()`? 

I tried running `python setup.py install` and got this error message: `django.core.management.base.CommandError: Can't find msgfmt. Make sure you have GNU gettext tools 0.15 or newer installed.` 

I resolved with [this SO post](http://stackoverflow.com/questions/35101850/cant-find-msguniq-make-sure-you-have-gnu-gettext-tools-0-15-or-newer-installed). Can you add a few lines to the readme that the `gettext` package is needed for manual setup?

allo- commented:
I wonder if you need the gettext binary or just python gettext packages. Can the CI test this?

allo- commented:
Probably there is a python gettext dependency in django anyway.

Another question is, if ``install_lib`` is the best install step for this. I had this from a SO answer, but it seems that the class has several other methods which can be used.

joshblum commented:
Rebasing would be great :) I don't have a strong opinion on the `install_lib` part, it seems to place the files in the correct location. What do you think?

allo- commented:
Err ... i used super the wrong way, but the other way it still fails, because ´´install_lib´´ is a old style class.

So either revert the commit and keep with the ``_install_lib.run(self)`` or try multiple inheritance from ``_install_lib`` and ``object`` (rather ugly, isn't it?)

allo- commented:
``install_lib`` isn't run i.e. for ``setup.py bdist_egg``. I think all binary packages should include prebuild ``.mo`` files.

joshblum commented:
Hmm let's just revert the commit then. What is the fix for `bdist_egg` then?

allo- commented:
Seems that neither ``bdist`` nor ``bdist_egg`` commands work as intended with ``install_lib``.

When i use ``build_py`` bdist works, but bdist_egg doesn't.

allo- commented:
The difference will be, if the target system of a binary distribution needs the ``gettext`` binary.

joshblum commented:
@allo- Thanks for the contribution and your patience during the review process!!

allo- commented:
Mh, i wanted to rebase it before a merge with less commits and a more readable commit message. 

If you like, push --force the revision before and i cleanup this branch later for a nicer merge.

And i am still not sure about the right command, but ``build_py`` will work until we find something better, which can be integrated then.

joshblum commented:
I squashed all of the commits in the end so it was just a single commit
with a combination of the other messages :)
El El mar, jul 12, 2016 a las 10:07, allo- <notifications@github.com>
escribió:

> Mh, i wanted to rebase it before a merge with less commits and a more
> readable commit message.
>
> If you like, push --force the revision before and i cleanup this branch
> later for a nicer merge.
>
> And i am still not sure about the right command, but build_py will work
> until we find something better, which can be integrated then.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/macropin/django-registration/pull/189#issuecomment-232058352>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABF01Lvr8ljd4bJF5_NVCloRh4Xuqu10ks5qU5-ugaJpZM4IzHLi>
> .
>

allo- commented:
I would have reworded it, for example the one commit and it's exact revert aren't important in the log.

allo- commented:
If possible, i would like you to use
https://github.com/allo-/django-registration/commit/df4a36b22126736872e5a269d0514ec541917db5
I not only wrote a more concise commit message, but also fixed the real
name field of the commit (which just had the username in the commits you
squashed).

joshblum commented:
@allo- checkout https://github.com/macropin/django-registration/commit/a1b31a087027502c0c989dfcbf658c617cb28f3c i rewrote the commit message

allo- commented:
ah, okay.

joshblum commented:
all good?

allo- commented:
The name Metadata is still just the nick, but just let it be. No big deal for a few lines.
